framework/src/samplers/Sampler.C

```diff
@@ -86,3 +116,38 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
 
   _generator.saveState();
 }
+
+void
+Sampler::setSampleNames(const std::vector<std::string> & names)
+{
+  _sample_names = names;
+
+  // Use assert because to check the size a getSamples call is required, which you don't
+  // want to do if you don't need it.
+  mooseAssert(getSamples().size() == _sample_names.size(),
+              "The number of sample names must match the number of samples returned.");
+}
+
+Sampler::Location
+Sampler::getLocation(unsigned int global_index)
+{
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  mooseAssert(_offsets.size() > 1,
+              "The getSamples method returned an empty vector, if you are seeing this you have "
+              "done something to bypass another assert in the 'getSamples' method that should "
+              "prevent this message.");
+
+  std::vector<unsigned int>::const_iterator iter =
+      std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```

Comment:
Why not std::lower_bound here? Then you wouldn't need a '- 1`.

Reply:
@aeslaughter - I'm still having a hard time with this. We routinely use lower_bound, so much so that I just put in a macro yesterday to replace the pattern of using lower bound to get the iterator to the right spot for insertions to occur (i.e. if the item is missing, it'll point where the item is supposed to go, if the item is present, it'll point at the item). Not less is used because that's the standard operator you need to define to have items stored in any sorted container (set or map). Greater than, is not normally required. I'm not convinced that this is correct because I think it misses edge cases. Can you provide a counter example?

Reply:
Lower bound usage in MOOSE: https://github.com/idaholab/moose/commit/6580d8254026826e6c0c08c78b2111612284c39b

Reply:
I added a unit test that shows the behavior I need. I do not need an insert location, I am looking for a starting index based on an offset. In the example below, given any global index in need the index that gives me the starting point of that subset.

```
GLOBAL:     0     1      2     3        4      5     6      7     8
VECTORS: [x_0, x_1, x_2, x_3], [y_0, y_1, y_2, y_3, y_4,...]
```

Reply:
Thanks, I see now.